metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala

```diff
@@ -0,0 +1,389 @@
+            .flatMap(_.iterator)
+            .filter {
+              case '\n' => false
+              case _ => true
+            }
+            .take(margin)
+          out.appendAll(truncatedLine)
+      }
+    }
+    out.toString()
+  }
+
+  private def evaluateWorksheet(
+      path: AbsolutePath,
+      token: CancelToken
+  ): Array[DecorationOptions] = {
+    val input = path.toInputFromBuffers(buffers)
+    // NOTE(olafurpg): the sbt dialects is the closest available syntax to worksheets.
+    val WorksheetDialect = dialects.Sbt1
+    val decorations = for {
+      ctx <- getContext(path)
+      source <- WorksheetDialect(input).parse[Source].toOption
+    } yield renderDecorations(path, ctx, source)
+    decorations.getOrElse(Array.empty)
+  }
+
+  private def renderDecorations(
+      path: AbsolutePath,
+      ctx: Context,
+      source: Source
+  ): Array[DecorationOptions] = {
+    val sectionInput = SectionInput(
+      path.toInputFromBuffers(buffers),
+      source,
+      Modifier.Default()
+    )
+    val sectionInputs = List(sectionInput)
+    val instrumented = Instrumenter.instrument(sectionInputs)
+    val rendered = MarkdownCompiler.buildDocument(
+      ctx.compiler,
+      ctx.reporter,
+      sectionInputs,
+      instrumented,
+      path.toString
+    )
+
+    val decorations = for {
+      section <- rendered.sections.iterator
+      statement <- section.section.statements
+    } yield renderDecoration(statement)
+
+    diagnostics.onPublishDiagnostics(
+      path,
+      reporter.diagnostics.toSeq,
+      isReset = true
+    )
+
+    decorations
+      .filterNot(_.renderOptions.after.contentText == commentHeader)
+      .toArray
+  }
+
+  private def renderDecoration(statement: Statement): DecorationOptions = {
+    val pos = statement.position
+    val range = new l.Range(
+      new l.Position(pos.startLine, pos.startColumn),
+      new l.Position(pos.endLine, pos.endColumn)
+    )
+    val margin = math.max(
+      20,
+      userConfig().worksheetScreenWidth - statement.position.endColumn
+    )
+    val isEmptyValue = isUnitType(statement) || statement.binders.isEmpty
+    val contentText = renderContentText(statement, margin, isEmptyValue)
+    val hoverMessage = renderHoverMessage(statement, margin, isEmptyValue)
+    DecorationOptions(
+      range,
+      new l.MarkedString("scala", hoverMessage),
+      ThemableDecorationInstanceRenderOptions(
+        after = ThemableDecorationAttachmentRenderOptions(
+          contentText,
+          color = "green",
+          fontStyle = "italic"
+        )
+      )
+    )
+  }
+
+  private def getContext(path: AbsolutePath): Option[Context] = {
+    for {
+      target <- buildTargets.inverseSources(path)
+      info <- buildTargets.scalaTarget(target)
+      scala <- info.info.asScalaBuildTarget
+      scalaVersion = scala.getScalaVersion
+      isSupported = ScalaVersions.isSupportedScalaVersion(scalaVersion)
```

Comment:
You mentioned worksheets work only in 2.12.10, but this is checking against all versions supported by Metals, isn't it?

Reply:
Great catch! Indeed. Changed the condition to check against the same Scala version